### ISSUE TITLE
How to avoid injecting new code into the asar archive file?

### ISSUE THREAD

**Issue by ironman0x7b2** [electron/electron#19671]
Hi,

I recently saw an article about injecting the malicious code into the asar archive file. Is there any solution to avoid this?

https://arstechnica.com/information-technology/2019/08/skype-slack-other-electron-based-apps-can-be-easily-backdoored/

**Comment by MarshallOfSound** [comment]
@ironman0x7b2 That article is unfortunately quite a lot of FUD.  Not going to invest a lot of time in this but here are some quick bullet points to hopefully put this to rest.

* If you modify an ASAR file it does actually fail the macOS code signing check if you try give that modified application to someone else.  The people who wrote this post fundamentally don't understand how Gatekeeper or security boundaries work so here is a quick overview.

> ASAR files are signed as part of the application bundle. The issue is that folks don't understand how gatekeeper works so let me try explain it here.

> When you download an application from the internet, macOS initially considers it "quarantined". When a quarantined application is first opened gatekeeper scans it _completely_ and if it's happy removes the quarantine tag and let's it launch.

> Once that quarantine tag is removed, gatekeeper will never run a complete check of that application again. Meaning the ASAR files are validated once, when the application is first launched.

> What people are seeing here is they're taking an application that gatekeeper has already signed off on, modifying it, and then asking why gatekeeper didn't stop them.

> If you took that modified application, zipped it up, uploaded it somewhere, downloaded it again and tried to run it, it would NOT work. Gatekeeper would boot that invalid application to the shadow realm.

* And most simply if you have write access (and sudo to modify `/Applications`) to a victims machine you have substantially more to worry about than injecting JS into an ASAR file.  You could replace the entire application, install a rootkit, install a self signed code signing certificate and trust it to bypass codesign checks, etc.

The TLDR is I stand firmly in the belief that this is not an actual issue, if it were the reporter would be collecting bug bounties from the many large Electron apps and would be getting a CVE assigned instead of writing a partially incorrectly and mostly FUD article about it.

Happy to discuss further how Gatekeeper / ASAR file integrity works but please remember to keep discussions civil folks 😄

**Comment by don-reba** [comment]
Why talk about Gatekeeper? This is not a Mac-specific vulnerability. In fact, the demo video shows Windows.

**Comment by sadreck** [comment]
Hi,

I believe this needs a bit more context as people seem to focus on specific aspects of the issue, therefore not realising the impact - perhaps I'm to blame for that.

If you have a look at github.com/ctxis/beemka there are videos for both linux and windows that prove that electron apps can be injected with JS code without modifying the binary signature. However this isn't a remote attack  - it's local. 

Here's a real life scenario that actually happened. I work as a security consultant and we once did a Red Team engagement for a client. The way we established persistence within the organisation was to inject a powershell payload within slack's electron.asar file. So every time slack would run on start up, we'd get reconnected to the internal network. 

Although this isn't a remote vulnerability to users, it's still a very real attack vector for malicious actors that target organisations (once they are already in).

Next week I'll upload a video where slack's cookies are being egressed, and another with VSCode sources.

The fix would be to simply add an integrity check, I'm not sure why there's such a pushback for this.

**Comment by MarshallOfSound** [comment]
@sadreck To avoid the inevitable miscommunications (that have already happened around this thing) can you send me an email and we'll try get on the same page in a more synchronous way.

**Comment by Kleissner** [comment]
Can somebody please suggest @MarshallOfSound for the Pwnie Award "Lamest Vendor Response"? Saying "FUD" while deflecting from the actual issue is quite a thing.

Unless the Electron developers are doing something about it, it is becoming the new Flash.

**Comment by sadreck** [comment]
@Kleissner come on mate, there's literally no reason to be like that. 

@MarshallOfSound sure thing, will do.

**Comment by danstur** [comment]
@MarshallOfSound As far as I can see Gatekeeper and /Application are Mac OS X specific protections which obviously won't work on Windows or Linux. 

It sounds like you're saying this is FUD because the exploit won't work on one of three operating systems.

**Comment by tonyganch** [comment]
@Kleissner, your comment is a violation of [the Electron Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md). You may consider this an official warning.

**Please do not interact with the project for 24 hours.** After that, please look through your open issues and edit them to ensure they're entirely on-topic.

**Comment by duanemoody** [comment]
Does the ASAR archive file need to be modifiable by the application at runtime after it's built, or is it possible to sign it at build time?

**Comment by jeroen7s** [comment]
Could someone reopen this please?
Seems like a pretty big vulnerability, especially with Password Management application like BitWarden using it. MacOS seems the least vulnerable, so talking about Gatekeeper as protection and closing the ticket seems to be deflecting from an outsiders perspective.

**Comment by PhilippLgh** [comment]
While I think that the article indeed creates FUD, I would also like to mention that for our Electron app ["Grid"](https://github.com/ethereum/grid) we are not using asar files but instead signed zip or tar packages.
Why?
- we have smaller / lighter updates for the UI portion of the app without having to go through full re-installation, and 60MB updates etc
- in this case signed updates are rather critical especially since we are dealing with crypto currencies

For the majority of applications an asar modification would break the application signature as described by @MarshallOfSound and therefore a re-distribution would not be possible or at least not as described. And as said before if the attacker has already root access to the computer you would have to worry about other things than them modifying your slack app.

Anyways, I believe that long-term the way we update our Electron applications will change (we don't re-install chrome to refresh websites either) and if you are interested how signed packages and electron can work together we've done a lot of open source work in this regard (unfortunately the docs need an update) and the packages are not production ready. But happy to discuss the reasoning and roadmap. If one is interested please check out:
https://github.com/PhilippLgh/electron-app-manager
https://github.com/PhilippLgh/ethereum-signed-packages

**Comment by bendemboski** [comment]
If we don't think there are any code changes or anything that make sense here, could we at least come up with a write-up/README about it? I have a lot of very security-conscious customers using my application *on windows*, so I'd really like writeup about the considerations on windows that is as clear as the one @MarshallOfSound provided for MacOS so that I have something to point to/tell any customers that might come across the arstechnica article.

**Comment by jeroen7s** [comment]
"therefore a re-distribution would not be possible or at least not as described."
This is not even the security risk here, on windows, most electron apps are installed in appdata, this would allow any malware without any admin rights to inject code easily into existing applications, without triggering any warnings

**Comment by PhilippLgh** [comment]
OK I see.. so the discussion centers more around the fact that a malware could do more harm on a local machine with less privileges. That's indeed a problem. A possible solution could be to write the asar checksum to a privileged space such as windows registry during installation and do an integrity check on startup. But I believe the whole reasoning behind using AppData was to avoid installer/updater elevation if possible. Another option would be to use signed packages which would also have the benefit that the asar (or zip or tar) file can be updated independently (see approach from my last comment). Which is probably the more secure, flexible and consistent solution but cannot be fixed by the Electron team unless the community establishes some PKI with root certs.

**Comment by tex0l** [comment]
I don't agree at all with the general argument conveyed in this discussion against fixing this which can be summed up by saying "anyone with that kind of access could do worse elsewhere". In native applications, we've invented code signing to prevent anyone from tampering with the application. We'd never invoke the argument saying "anyone with that kind of access could do worse elsewhere" to avoid code signing today. Ensuring integrity of the code contained is the ASAR is the exactly the same, it has the same privileges as the binary interpreting the code which is signed, and therefore should be handled with the same caution.

I don't know how, where in the stack or with what features of the OS, but having a way to ensure code integrity after it's been installed is critical.
